Style, lint, more throttling, more logging, more options

sctice opened the pull request:
This captures all of the changes that I made in order to run a recent backfill of a few days over which we'd failed to capture orders:

- Ran it through Black
- Updated the Dockerfile and requirements.txt to get it building again
- Ran it through pylint, fixed some complaints, and added more logging
- Updated the Dockerfile to support linting within the container
- Addressed the rest of the linter complaints
- Fixed a bug when `ListOrders` returns a single order
- Added an explicit delay between `ListOrderItems` requests to avoid depleting our quota
- Added support for `NextToken` on `ListOrderItems` (but I don't think we've ever needed it)
- Added support for explicit start and end dates as an alternative to `DAYS_AGO`

Apologies. I know that's a lot of changes.

Connects iFixit/ifixit#34656

CC @iFixit/bi

sctice-ifixit committed:
.gitignore: ignore .env files

sctice-ifixit committed:
*.py: reformat with Black

sctice-ifixit committed:
Dockerfile: update to official python image

At some point, `apk add python` stopped also installing pip3, so
building the image with the current Dockerfile was broken. Instead,
let's use a more stable environment that sets up python and pip for us
and lets us call them as just `python` and `pip`. We choose Python 3.8
because that's the latest version that the `mws` package supports, and
to make sure that our version of `mws` works with 3.8, we update `mws`
to the latest stable version.

Additionally, move the Dockerfile around so that we're making better use
of caching.

sctice-ifixit committed:
main: throttling, more logging, misc errors

This is a big one. Sorry.

Change the throttling strategy to wait long enough between requests so
that we only use up our quota slowly. In the event that we exhaust the
quota, wait long enough to refill our pool substantially and then
resume. Because most of our time is now spent sleeping between
`ListOrderItem` requests, we shouldn't really need to wait between
`ListOrders` requests, and it's unlikely that we'll run into throttling
there.

Add logging of the happy path throughout so that it's easier to see
what's happening. When we encounter an error, don't assume that it's due
to rate limiting and instead print out the entire error (it gets printed
out as XML).

Handle the case that the list of orders contains exactly one order in
the same way that we handle a list of exactly one item (the library
parses the XML into a dict rather than a list containing a dict).

Exchange some maps for a simple loop over the orders generator; note
that since we're iterating orders, we fetch 100 orders, then fill in the
items for them, then fetch up to another 100 orders, and so on until we
run out of orders.

Use `isinstance()` instead of `type() ==` because `pylint` prefers the
former.

Rename some methods and partials to be more intuitive (to me).

sctice-ifixit committed:
Makefile

Add a rudimentary Makefile so that it's a bit easier to keep the style
and linting up to date. Because we're running the linter outside of the
container, it's missing all of the requirements that we install within
the container. I don't think it should be too difficult to install
`pylint` in the container and run it there, but I didn't want to do it
right now.

Pylint is still complaining about some missing docblocks in addition to
unused imports, but no major complaints.

sctice-ifixit committed:
dev: lint in container

I want to drive the lint down to zero so that real issues stand out, and
part of that is getting it to run where we have all of the requirements.

sctice-ifixit committed:
lint: fix import errors

pylint needs to see this directory as a module in order to know that it
can make relative imports.

sctice-ifixit committed:
lint: fix naming errors

For reasons I don't entirely understand, pylint sees some of the
declarations at the top of `__main__.py` (but not, e.g., `orders_api`)
as constants, and it wants them to be uppercase. I think that's an
improvement because it makes them stand out as environment arguments.

It also doesn't like `fn` as a variable name because it's too short.
I agree less on this one, but it's easier to choose a more specific name
than to configure pylint for the exception.

sctice-ifixit committed:
lint: add documentation

Also:

- Replace set_document_id with a simple operation.
- Rename orders_generator to start with a verb.
- Rename add_order_items to more clearly indicate what the method
  actually does on top of set_order_items.

sctice-ifixit committed:
ListOrderItems: support for NextToken

I don't know whether we ever have orders with more than the maximum
number of items for one request (the docs don't even indicate what that
maximum is), but now we _should_ support such large orders.
Unfortunately, it's difficult to test because I don't currently have
access to an order with enough items such that NextToken is present. At
the very least, we should see this task fail if the logic is flawed
_and_ we run into such an order, and then we can fix it because we'll
have an actual test case.

sctice-ifixit committed:
ListOrderItems: bump delay between requests

I found that waiting 1.5s between ListOrderItems requests was depleting
our quota after a few hundred requests (less than one day's worth of
orders in the US/EU). A little math based on the number of requests
between rate limit messages indicates that waiting 1.75s brings us
closer to the 2s target. With the new value, I saw no rate limit
messages on a larger backfill of several thousand orders.

sctice-ifixit committed:
Support explicit start and end dates

I found myself in a situation where I needed to backfill a one-day hole
more than 10 days in the past, and I got tired of waiting for the
backfill on all of the days that I didn't care about. Now if you specify
an explicit START_DATE, it overrides DAYS_AGO. If you specify END_DATE,
then we honor that; otherwise, we assume END_DATE is now.

sctice commented:
Going commit by commit should explain all of the changes. The big thing that started me down this path was that the image would no longer build because of changes to the python package the old version used to install (we were no longer getting `pip3` for free). From there, I wanted to modernize and add functionality to backfill just a specific hole because I found that it takes quite a while to import all of the orders and their items for an entire day.

djmetzle commented on Dockerfile:
```diff
@@ -1,9 +1,11 @@
+RUN pip install --no-cache-dir -r requirements.txt
 
-ENTRYPOINT python3 -m amazon-mws-analytics
+COPY *.py ./
```
```suggestion
COPY *.py .
```
You don't need the trailing slash i don't think...

sctice replied:
From [the docs](https://docs.docker.com/engine/reference/builder/#copy):

> If multiple `<src>` resources are specified, either directly or due to the use of a wildcard, then `<dest>` must be a directory, and it must end with a slash `/`.

djmetzle commented on Dockerfile:
```diff
@@ -1,9 +1,11 @@
+COPY *.py ./
+
+WORKDIR /
+CMD ["python", "-m", "app"]
```
Your Docker-fu is getting strong @sctice!

I'm little off-put by the `WORKDIR /` though. The `-m` flag implicitly running `__main__.py` surprises me. Not that it's wrong, its just a little surprising.

sctice replied:
I hear you. I just copied the previous invocation because it works and using relative imports was breaking in unexpected ways when I tried to run the main script directly. I suspect that might not be an issue after creating `__init__.py`, but I'm not positive.

djmetzle commented on Dockerfile:
```diff
@@ -1,11 +1,15 @@
+CMD ["pylint", "app"]
 
+FROM base AS prod
+COPY *.py ./
```
This is duplicated now. Did you intend that? Does pylint mutate the sources somehow?

sctice replied:
It's intended. Because of the `FROM base AS prod`, we're starting from the `base` image rather than the `lint` image, so we don't have the `*.py` files that we copied in there (nor `pylint` and its dependencies).

djmetzle commented on __main__.py:
```diff
@@ -12,28 +12,40 @@
+    if "END_DATE" in os.environ:
+        END_DATE = iso8601.parse_date(os.environ["END_DATE"]).isoformat()
+else:
+    DAYS_AGO = 1 if "DAYS_AGO" not in os.environ else int(os.environ["DAYS_AGO"])
```
Gah. That's ugly. Does python have some kind of ternary or short-circuiting to leverage here?

sctice replied:
You're _looking_ at Python's ternary, which is short-circuiting. In case it isn't clear, Guido doesn't want you to use it.

```
In [1]: 1 if True else BOOM()
Out[1]: 1
```

djmetzle commented:
Bunch of pesky comments maybe worth reviewing, but this looks good. Awesome stuff @sctice!

CR :package:

andyg0808 commented on __main__.py:
```diff
@@ -1,11 +1,12 @@
 
-from .orders import get_order_items, orders_generator, set_document_id
+from .orders import set_order_items, orders_generator, set_document_id
 from .utils import make_ratelimit_aware
```
The `isort` tool will autosort these for us, and follows some kind of convention, as far as I know.

andyg0808 commented on utils.py:
```diff
@@ -55,12 +55,12 @@ def convert_types(key, value):
-        except error:
-            print("ratelimited", file=sys.stderr)
+        except error_cls as err:
+            print(f"error: {str(err)}", file=sys.stderr)
```
Does 
```suggestion
            print(f"error: {err}", file=sys.stderr)
```
work?

sctice replied:
Bah. It certainly _should_ (it does). I'll fix it.

andyg0808 commented on Makefile:
```diff
@@ -0,0 +1,13 @@
+
+.PHONY: style
+style:
+	docker run -it --rm -v "$(PWD)":/code jbbarth/black *.py
```
This image looks pretty out-of-date; it's two years old, and it looks like `black` releases more often than that: https://pypi.org/project/black/#history

Also, I'm obliged to yell about depending on random people's packages or Docker images. I don't like trusting it.

Maybe we could do the same thing we do below for the `lint` case (I don't know if this entrypoint will work for sure):
```suggestion
	docker build --target tools -t $(REPO):lint .
	docker run --rm -it --entrypoint="black" $(REPO):lint 
```

sctice replied:
Yeah. I was just being lazy. I agree with all of your points, and I'll update it.

andyg0808 commented on Dockerfile:
```diff
@@ -1,11 +1,15 @@
 RUN pip install --no-cache-dir -r requirements.txt
 
+FROM base AS lint
+RUN pip install --no-cache-dir pylint
```
```suggestion
FROM base AS tools
RUN pip install --no-cache-dir pylint black
```

andyg0808 commented on Makefile:
```diff
@@ -2,12 +2,13 @@ REPO=amazon-mws-analytics
 .PHONY: lint
 lint:
-	docker run --rm -it -v "$(PWD)":/data cytopia/pylint *.py
+	docker build --target lint -t $(REPO):lint .
```
To go with the other suggestions:
```suggestion
	docker build --target tools -t $(REPO):lint .
```

andyg0808 commented:
CR :+1: through 5c0f281
I have some comments, but nothing blocking.
deploy_block :+1: on comments.

djmetzle commented:
Yeah, +1 vote on not using an out of date black container. Looks easy enough to explicitly install it with pip.

sctice-ifixit committed:
utils: Stop explicitly casting err

We're interpolating it into a string. Python knows that it needs a
string.

sctice-ifixit committed:
black: stop depending on rando image

We can just as well install it ourselves as part of the "dev" image.
While I was making this change, I realized that we can just mount and
lint the working directory directly, in which case we don't need to copy
our source files into the image.

Note that we had to do more work than expected to install black
ourselves. There's no Alpine wheel for a black dependency that needs
compilation, so we end up having to compile it ourselves. The
`--virtual` and `apk del` business lets us install a "virtual" package
for the black dependencies and then delete that package after installing
black to keep the image size down (since we don't actually need the
build environment).

See https://github.com/psf/black/issues/1106
See https://stackoverflow.com/questions/46221063/what-is-build-deps-for-apk-add-virtual-command/46222036

sctice-ifixit committed:
Add isort and run it

We now sort our imports before we run black so that they're always in a
consistent order and formatted the way black likes.

sctice commented:
It turned out to be a _little_ more complicated than expected to install `black` ourselves, but I think it's a big improvement.

I also set up `isort` for sorting our imports and ran everything through the new `style` and `lint` make targets.

Finally, I changed the `lint` make target to work directly on the working directory mounted in instead of python files baked into the image. Now we don't have to keep rebuilding the `dev` image.

un_deploy_block 🎏

andyg0808 commented on Dockerfile:
```diff
@@ -3,11 +3,13 @@ WORKDIR /app
-RUN pip install --no-cache-dir pylint
-COPY *.py ./
+FROM base AS dev
+# Alas, black depends on regex, which doesn't have an Alpine wheel, which means
+# we need to compile it, which means we need a build environment.
+RUN apk add --no-cache --virtual .black-build-deps gcc musl-dev \
+   && pip install --prefer-binary --no-cache-dir pylint black isort \
+   && apk del --no-cache .black-build-deps
```
Thank you!
I feel like the only appropriate response to this level of headache is to award you one Austrian Yak for your troubles:
![Yak](https://upload.wikimedia.org/wikipedia/commons/thumb/3/32/Yak_Kailash.JPG/273px-Yak_Kailash.JPG)

sctice replied:
I really do think it's the right way to go, the errors were clear, and the internet had my back.

andyg0808 commented:
CR :+1: through d011120

djmetzle commented:
CR :gear:

sctice-ifixit committed:
isoformat() of datetime, not timedelta

Whoops! I encountered this bug while re-testing the "DAYS_AGO"
functionality. Testing the fix also brought to my attention that this
container is running in UTC, so `datetime.now()` gives a UTC time, but
I'm pretty sure that isn't new with the recent changes.

sctice commented:
Whoops. I found a bug while I was re-testing, and testing the fix also brought to my attention that the container is running in UTC. I don't believe that's new behavior (although I'd appreciate another opinion on that), but is it the behavior that we want? I don't think it's a problem since we're passing an iso8601-formatted timestamp, which includes a timezone delta (or is implicitly UTC without one), but it can make the logs slightly confusing if you're expecting Phoenix times.

QA 🎏  on having tested now with both `DAYS_AGO` and `START_DATE` and optional `END_DATE`.

andyg0808 commented:
@sctice I think you're right that it was already running in UTC. I'm inclined to believe that it's working correctly, since it seems like the data is matching what folks expect, and I'm loath to risk something breaking if we change it.

CR :+1: through c629ad0

djmetzle commented:
That's kewl. CR :clock7: 

I would suggest setting the container localtime pointer to Phoenix, but not if the rest of the date logic is assuming Zulu..
